Remove and Set Filters (#11)
overture-stack/sqon-builder: README.md

## API

### SQONBuilder default export
`SQONBuilder(sqon: SQONBuilder | SQON | string)`

The package default export is the `SQONBuilder`. This is a function that will generate a `SQONBuilder` object from another `SQONBuilder`, a `SQON` object, or JSON `string`.

> Note: This will attempt to parse the provided string as JSON and will then validate that the contents are a valid SQON. If the provided string cannot be parsed from JSON a `SyntaxError` will be thrown. If the parsed string is not a valid SQON, a [`ZodError`](https://zod.dev/ERROR_HANDLING) will be thrown.

Example: `const builder: SQONBuilder = SQONBuilder({op: 'in', content: {fieldName: 'name', value: ['Jim']}});`

> The `SQONBuilder` function also provides many static functions that can be used to generate a new SQONBuilder without having a SQON to start with.

Reviewer: ```suggestion
The `SQONBuilder` function also provides many static methods that can be used to generate a new SQONBuilder without having a SQON to start with.
```

Example: `const builder: SQONBuilder = SQONBuilder.in('name', 'Jim');`

> The `SQONBuilder` object that is returned stores the value of the generated SQON which can be accessed as a string with `.toString()` or as an object with `.toValue()`. From the `SQONBuilder` object you can now combine the stored SQON value with other filters or combination operations.

Reviewer: ```suggestion
The returned `SQONBuilder` object stores the value of the generated SQON which can be accessed as a string with `.toString()` or as an object with `.toValue()`. From the `SQONBuilder` object you can now combine the stored SQON value with other filters or combination operations.
```

Example: `builder.lt('age','30').gt('score',95);`

#### Filter: In

> `SQONBuilder.in(fieldName: string, value: ArrayFilterValue) => SQONBuilder`

Reviewer: this syntax seems a bit confusing (looks like an arrow function).
maybe I need some help reading it correctly?

Reviewer: ah, reading the sentence bellow maybe this clears the example a little?
```suggestion
`SQONBuilder.in(fieldName: string, value: ArrayFilterValue)` returns a `SQONBuilder`
```

Creates a filter requiring the given field to have one of the given values. The value matches the `ArrayFilterValue` type which representes a string, a number, or an array of strings and numbers.

Example: `SQONBuilder.in('name',['Jim','Bob'])`

#### Filter: Greater Than

> `SQONBuilder.gt(fieldName: string, value: number) => SQONBuilder`

Reviewer: if you like the change above, adding this for consistency
```suggestion
`SQONBuilder.gt(fieldName: string, value: number)` returns a `SQONBuilder`
```

> Greater Than operator. Create a filter requiring the given field to be greater than the given value.

Reviewer: ```suggestion
"Greater Than" operator (i.e. `>`). Create a filter requiring the given field to be greater than the given value.
```

Example: `SQONBuilder.gt('age',21)`

#### Filter: Lesser Than

> `SQONBuilder.lt(fieldName: string, value: number) => SQONBuilder`

Reviewer: ```suggestion
`SQONBuilder.lt(fieldName: string, value: number)` returns a `SQONBuilder`
```

> Lesser Than operator. Create a filter requiring the given field to be lesser than the given value.

Reviewer: ```suggestion
"Lesser Than" operator (i.e. `<`). Create a filter requiring the given field to be lesser than the given value.
```

Example: `SQONBuilder.lt('count', 100)`

#### Combine: And

> `SQONBuilder.and(sqon: SQON | SQON[]) => SQONBuilder`

Reviewer: ```suggestion
`SQONBuilder.and(sqon: SQON | SQON[])` returns a `SQONBuilder`
```

All filters in the resulting SQON must be true.

Example: `SQONBuilder.and( [someSqon, anotherSqon] )`

#### Combine: Or

> `SQONBuilder.or(sqon: SQON | SQON[]) => SQONBuilder`

Reviewer: ```suggestion
`SQONBuilder.or(sqon: SQON | SQON[])` returns a `SQONBuilder`
```

At least one filter in the resulting SQON must be true.

Example: `SQONBuilder.or( [someSqon, anotherSqon] )`

#### Combine: Not

> `SQONBuilder.not(sqon: SQON | SQON[]) => SQONBuilder`

Reviewer: ```suggestion
`SQONBuilder.not(sqon: SQON | SQON[])` in a `SQONBuilder`
```

None of the filters in the resulting SQON can be true.

Example: `SQONBuilder.not( [someSqon] )`

#### From
`SQONBuilder.from(input: unknown) => SQONBuilder`

Build a new SQON from a string or from a JSON object.

Example with string:

```ts
SQONBuilder.from(
  '{"op":"and","content":[{"op":"in","content":{"fieldName":"name","value":"Tim"}},{"op":"gt","content":{"fieldName":"age","value":"19"}}]}',
);
```

> This differs from the default function because it will accept any input. The use case for this function is when you have a JS object that could potentially be a valid SQON, you can pass this `uncheckedObject` to `SQONBuilder.from(uncheckedObject)` and validation will be performed before the `SQONBuilder` is returned, if possible.

Reviewer: This says .from behaves differently, but based on the description for SQONBuilder itself, I'm not 100% sure I understand what the difference is here.

let me see if I can spell out what I see: both take SQON objects, or a string, or a SQONBuilder; both validate and respond the same way, e.g. SQONBuilder, or SyntaxError or ZodError.

Perhaps the difference is strictly on the input typing? if so, I think that could be explicitly stated here, for clarity. I've added "unknown" in case that's right. Feel free to edit this "suggestion comment" to your convenience.

```suggestion
This differs from the default function in that it will accept any input (i.e. type `unknown`). The use case for this function is when you have a JS object that could potentially be a valid SQON, you can pass this `uncheckedObject` to `SQONBuilder.from(uncheckedObject)` and validation will be performed before the `SQONBuilder` is returned, if possible.
```

If the provided string cannot be parsed from JSON a `SyntaxError` will be thrown.

If the provided object or parsed string is not a valid SQON, a [`ZodError`](https://zod.dev/ERROR_HANDLING) will be thrown.

### SQONBuilder Object

The `SQONBuilder` type is an object that stores a `SQON` object and exposes functions to modify that `SQON`, returning a new `SQONBuilder` object. This allows a user to chain function calls together to create complex filters one step at a time.

All filter and combination methods listed above can be called from the returned `SQONBuilder` object:
* In: `.in(fieldName: string, value: ArrayFilterValue) => SQONBuilder`
* GreaterThan: `.gt(fieldName: string, value: number) => SQONBuilder`
* LesserThan: `.lt(fieldName: string, value: number) => SQONBuilder`
* And: `.and(sqon: SQONBuilder | SQON) => SQONBuilder`
* Or: `.in(sqon: SQONBuilder | SQON) => SQONBuilder`
* Not: `.in(sqon: SQONBuilder | SQON) => SQONBuilder`

Example: `const mySqon: SQON = SQONBuilder.in('name', ['Jim']).or(SQONBuilder.gt('age', 25).lt('score', 50)).toValue();`

> All builder methods are side-effect free: modifications made with to a `SQONBuilder` are additive on top of a cloned SQON from the previous builder, so that the original builder is not modified. This can let you, for example, create a builder as a base SQON to make variations on in your application:

Reviewer: 🎉 function purity ftw

```ts
const base = SQONBuilder.in('name', ['Jim']);
const ageRestricted = base.gt('age', 25);
// {"op":"and","content":[{"op":"in","content":{"fieldName":"name","value":["Jim"]}},{"op":"gt","content":{"fieldName":"age","value":25}}]}

const scoreRestricted = base.lt('score', 50);
// {"op":"and","content":[{"op":"in","content":{"fieldName":"name","value":["Jim"]}},{"op":"lt","content":{"fieldName":"score","value":50}}]}
```

As filters and combinations are added to a `SQONBuilder`, the `SQON` maintained will be [reduced](#reduce-sqon), if possible. This will result in matching filters and combinations to be combined, and empty combinations to be removed.

In addition to the filter and combination functions, there are some additional methods that can modify the filters stored in the existing SQON, and then two methods for outputing the stored `SQON` as either a string with [`.toString()`](#tostring) or a value with [`.toValue()`](#tovalue)

#### ToString

> `builder.toString() => string`

Reviewer: renamed builder for consistency
```suggestion
`SQONBuilder.toString()` returns a `string`
```

Author: Need better clarity on this document of outlining the member methods of the default export vs the members on a SQONBuilder object...

In the examples I provided, the `SQONBuilder` token specifically refers to the default export, but most of its methods return an object of __type__ `SQONBuilder`, hence the confusion. For `toString` and `toValue`, that function is not a member of the default export: ie. `SQONBuilder.toString()` is not a function. However, a previously created builder has these methods: `SQONBuilder(mySqon).toString()`.

This is all to say I can't accept this change as is, but some overhauling of the whole readme is needed to clarify this.

Reviewer: I see! that makes sense now.
perhaps we need different "labels" for each of those.

e.g. base "builder function" vs "built sqons". 
going to look into jQuery named things, because this implementation reminds me a lot of it

Return stored `SQON` value as string.

#### ToValue

> `builder.toValue() => SQON`

Reviewer: ```suggestion
`SQONBuilder.toValue()` returns a `SQON`
```

Return stored `SQON` value as an object.

#### Remove Exact Filter

> `.removeExactFilter(filter: FilterOperator) => SQONBuilder`

Reviewer: ```suggestion
`SQONBuilder.removeExactFilter(filter: FilterOperator)` returns a `SQONBuilder`
```

Find exact matching filter and remove it from the SQON.

For filters with an array of values, the order of the array will be ignored during matching.

> Note: This only looks for filters at the root of the sqon or in the content of the top level combination operator.
This will not search recursively through the SQON.

```ts
const initial = SQONBuilder.in('name', 'Jim').gt('score', 50);
initial.removeFilter({op: FilterKeys.In, content: {fieldName: 'name', value: ['Jim']}});
// {op: 'gt', content: {fieldName: 'score', value: 50}}
```

#### Remove Filter

> `.(fieldName: string, op?: FilterKey, value?: FilterValue) => SQONBuilder`

Reviewer: Think you forgot the method in this example
```suggestion
`SQONBuilder.removeFilter(fieldName: string, op?: FilterKey, value?: FilterValue)` returns a `SQONBuilder`
```

If a filter is found that matches the fieldName and op, then all matching values form the provided array will
be removed from the filter. This lets you remove select values from an array filter without removing the entire
filter.

> Note: This only looks for filters at the root of the sqon or in the content of the top level combination operator.
This will not search recursively through the SQON.

_Example removing all filters for a property:_
```ts
const builder = SQONBuilder.in('name', ['Jim', 'Bob']).gt('age', 20).lt('age', 50);
// {"op":"and","content":[{"op":"in","content":{"fieldName":"name","value":["Jim","Bob"]}},{"op":"gt","content":{"fieldName":"age","value":20}},{"op":"lt","content":{"fieldName":"age","value":50}}]}

// Remove all filters on 'age'
builder.removeFilter('age');
// {"op":"in","content":{"fieldName":"name","value":["Jim","Bob"]}}
```

_Example removing some values from an array filter:_
```ts
const builder = SQONBuilder.in('name', ['Jim', 'Bob', 'May']);
builder.removeFilter('name', FilterKeys.In, ['Jim', 'Bob', 'Sue']);
// {"op":"in","content":{"fieldName":"name","value":["May"]}}
```

#### Set Filter

> `.setFilter(fieldName: string, op: FilterKey, value: FilterValueMap[FilterKey]) => SQONBuilder`

Reviewer: ```suggestion
`SQONBuilder.setFilter(fieldName: string, op: FilterKey, value: FilterValueMap[FilterKey])` returns a `SQONBuilder`
```

Add a specific filter to the content of the top level operator, or replace a matching filter (same `op` and `fieldName`) with the new value specified.

If the current SQON is just a filter, then the existing filter and this new filter will be combined with an `and` operator.

Example:
```ts
const builder = SQONBuilder.gt('age', 50).in('name', ['Jim', 'Bob']);
builder.setFilter('name', FilterKeys.In, ['Jim', 'Sue']);
// {"op":"and","content":[{"op":"gt","content":{"fieldName":"age","value":50}},{"op":"in","content":{"fieldName":"name","value":["Jim","Sue"]}}]}
```

### Reduce SQON

> `reduceSQON(sqon: SQON) => SQON`

Reviewer: ```suggestion
reduceSQON(sqon: SQON)` returns a `SQON`
```

The `reduceSQON` function is used internally by the SQON builder to reduce the complexity of a SQON by removing redundant operators and collecting similar filters. For example, if there is an `and` combination wrapping a single filter, the `and` operator can be removed and replaced by the single filter.

The reducer function is exported so you are able to run the reduction algorithm on SQONs independent of the SQONBuilder.

```ts
import { reduceSQON } from '@overture-stack/sqon-builder`;

const sqon: SQON = { op: 'and', 'content': [ { op: 'lt', content: { fieldName: 'age', value: 100 } } ] };

const reduced = reduceSQON(sqon);
// { op: 'lt', content: { fieldName: 'age', value: 100 } }
```

### Check Matching Filter

`checkMatchingFilter(a: FilterOperator, b: FilterOperator) => boolean`

The `checkMatchingFilter` function will compare two `FilterOperators` and return is they match in all properties. This ensures they have the same filter operation (`op` value), same `fieldName` and matching `value`. Note that for array values, the match is performed independent of order and after removing duplicates - it is a match on the logical content not on the exact array.

```ts

> import { checkMatchingFilter } from '@overture-stack/sqon-builder`;

Reviewer: To keep it consistent with other examples
```suggestion
import { checkMatchingFilter } from '@overture-stack/sqon-builder';

```

const filterA = { op: FilterKeys.In, content: { fieldName: 'name', value: ['Jim', 'Bob', 'May'] } };
const filterB = {	op: FilterKeys.In, content: { fieldName: 'name', value: ['May', 'Jim', 'Bob'] } };

const matchResult = checkMatchingFilter(filterA, filterB); // true
```

> ## Types and SQON Validation

Reviewer: it occurs to me all this conceptual info can be moved before the methods, etc.

Author: Definitely needs a dedicated editing pass over the whole README.

### SQON Type Composition

A SQON is composed from a series of nested `Operators` that take the general structure:

```json
{
	"op": "<operation code>",
	"content": <Value or array of values>
}
```

There are two types of Operators:

> 1. `FilterOperators` - Define a rule used to filter data based on matches to a specific field., for example the `in` filter will find all data that has a field value matching one of the included values in the filter.

Reviewer: ```suggestion
1. `FilterOperators` - Define a rule used to filter data based on matches to a specific field. For example, the `in` filter will find all data that has a field value matching one of the included values in the filter.
```

2. `CombinationOperators` - Combine multiple operators with boolean logic. 

#### FilterOperators

The `content` property of a `FilterOperator` is a single object of the form:

```json
{
	"fieldName": "<Field name to filter on>",
	"value": <Value or array of values>,
}
```

> Some filters accept an array of values, number or string, and some only accept a single number. These are differentiated in this library by the types `ArrayFilter` and `ScalarFilter`. You also have access to all the corresponding operation keys in the `FilterKeys`, `ArrayFilterKeys`, and `ScalarFilterKeys` records.

Reviewer: ```suggestion
There are two categories of filters that differ based on the types of values they accept as an argument:
1. `ArrayFilter` - accept a number, a string, or an array of numbers or strings.
2. `ScalarFilter` - accept only a single number
Types are exported that represent these filters, their expected value types `ArrayFilterValue`/`ScalarFilterValue`, as well as types that define the keys available for each filter type and all filters together: `ArrayFilterKeys`, `ScalarFilterKeys` and  `FilterKeys`.
```
